modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/GridCacheLockImpl2Unfair.java

```diff
@@ -0,0 +1,784 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.datastructures;
+
+import java.io.Externalizable;
+import java.nio.ByteBuffer;
+import java.util.ArrayDeque;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import javax.cache.processor.EntryProcessorResult;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.IgniteException;
+import org.apache.ignite.IgniteInterruptedException;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.IgniteInternalFuture;
+import org.apache.ignite.internal.processors.cache.GridCacheContext;
+import org.apache.ignite.internal.processors.cache.GridCacheIdMessage;
+import org.apache.ignite.internal.processors.cache.IgniteInternalCache;
+import org.apache.ignite.internal.util.typedef.internal.S;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.lang.IgniteInClosure;
+import org.apache.ignite.plugin.extensions.communication.MessageReader;
+import org.apache.ignite.plugin.extensions.communication.MessageWriter;
+
+import static org.apache.ignite.internal.managers.communication.GridIoPolicy.P2P_POOL;
+
+/**
+ * Unfair implementation of reentrant lock based on IgniteCache.invoke.
+ */
+public final class GridCacheLockImpl2Unfair extends GridCacheLockEx2 {
+    /** Reentrant lock name. */
+    private final String name;
+
+    /** Cache context. */
+    private final GridCacheContext<GridCacheInternalKey, GridCacheLockState2Base<UUID>> ctx;
+
+    /** Internal synchronization object. */
+    private final LocalSync sync;
+
+    /** Key for shared lock state. */
+    private final GridCacheInternalKey key;
+
+    /** ReleasedMessage handler. */
+    private final IgniteInClosure<GridCacheIdMessage> releaser;
+
+    /**
+     * Constructor.
+     *
+     * @param name Reentrant lock name.
+     * @param key Reentrant lock key.
+     * @param lockView Reentrant lock projection.
+     */
+    public GridCacheLockImpl2Unfair(String name, GridCacheInternalKey key,
+        IgniteInternalCache<GridCacheInternalKey, GridCacheLockState2Base<UUID>> lockView) {
+        assert name != null;
+        assert key != null;
+        assert lockView != null;
+
+        this.name = name;
+        this.key = key;
+        this.ctx = lockView.context();
+
+        // final for passing into anonymous classes.
```

Comment:
do we need such comment here? its just java spec

Reply:
But for Java 8 final is necessary. So in future, we can remove final. And that comment explains it for the future generation of developers.

Reply:
fixed (remove)